option/option.go

```diff
@@ -0,0 +1,69 @@
+package option
+
+import (
+	"github.com/repeale/fp-go"
+)
+
+type Option[T any] struct {
+	value    T
+	hasValue bool
+}
+
+func Some[T any](value T) Option[T] {
+	return Option[T]{value, true}
+}
+
+func None[T any]() Option[T] {
+	return Option[T]{}
+}
+
+func IsSome[T any](option Option[T]) bool {
+	return option.hasValue
+}
+
+func IsNone[T any](option Option[T]) bool {
+	return !option.hasValue
+}
+
+func GetOrElse[T any](onNone fp.Lazy[T]) func(Option[T]) T {
+	return func(option Option[T]) T {
+
+		if IsSome(option) {
+			return option.value
+		}
+
+		return onNone()
+	}
+}
+
+func Match[T any](onNone fp.Lazy[T], onSome fp.LazyVal[T]) func(Option[T]) T {
+	return func(option Option[T]) T {
+
+		if IsSome(option) {
+			return onSome(option.value)
+		}
+
+		return onNone()
+	}
+}
+
+func Map[T any](fn fp.LazyVal[T]) func(o Option[T]) Option[T] {
```

Comment:
Did you though about having something to squash the structure, like if the return type of `fn` is `Option[C]` the final return type of the map would be `Option[Option[C]]` and having something to squash the types into just `Option[C]` could be useful.

Reply:
I think you are looking for [fp.Chain](https://github.com/repeale/fp-go/pull/1/files/5253dd2bf5de28c663dded419fea20412cd7252c#diff-1ac97a144b121797df02eb5cc83716df12d40eec7cf22a7f67bc2146f7b69b43R61).
Maybe this could be improved but should solve your issue.
The signature does not force you to work on the same type but can return a generic `Option[B]`